### ISSUE TITLE
Puma ~> 5.5.1 returns 400 when User-Agent contains ASCII 31 (Unit Separator)

### ISSUE THREAD

**Issue by iainbethune** [puma/puma#2763]
**Describe the bug**
Since puma 5.5.1 (specifically the fix for https://github.com/puma/puma/security/advisories/GHSA-48w2-rm65-62xx), puma returns a 400 response to any requests with a User-Agent string containing the ASCII 31 character.  The following is logged:

```
HTTP parse error, malformed request ("GET /" - (-)): #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>
```

This is breaking our legacy application which includes `\x1f` as a separator in its user-agent string. 

**Puma config:**

```
threads 1, (ENV['MAX_THREADS'] || 5)
workers 2
```

**To Reproduce**
Using the Hello World example, with the above config

```
run lambda { |env| [200, {"Content-Type" => "text/plain"}, ["Hello World"]] }
```

Run it with:

```
bundle exec puma -C config.rb hello.ru
```

Run a GET with the user-agent set to "user agent":

```
curl -v -A "user agent" http://localhost:9292
```

Returns 200 OK and "Hello World" as expected.

Run a GET with the user-agent set to "user\x1fagent":

```
curl -v -A $(echo -ne "user\x1fagent") http://localhost:9292
```
Observe the 400 response and server logs:

```
HTTP parse error, malformed request ("GET /" - (-)): #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>
```

**Expected behavior**
This is believed to be a valid user-agent, and puma should return a 200 response.  For reference other HTTP servers, such as nginx, Python SimpleHTTPServer, although apache does return 400 and "found an invalid character in header value"

**Desktop (please complete the following information):**
 - OS: Mac OS 12.0.1
 - Puma Version 5.5.1, 5.5.2

**Comment by iainbethune** [comment]
Appears to be a dupe of https://github.com/puma/puma/issues/2722 - sorry

**Comment by iainbethune** [comment]
Actually, re-opening as I think this should have been resolved in 5.5.2 but it appears not to be the case.

**Comment by nateberkopec** [comment]
[The RFC](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.4) is pretty clear that in the ASCII character set, only visible characters are allowed. Allowing control characters leads to bugs and security holes, like the one we had to fix in this case.

Accepting non-visible ASCII characters was a bug, so it was fixed in a patch release.
